### PR TITLE
Mutual ratls example

### LINKED ISSUE #1: Creating Gramine-based mutual RA-TLS example
- [ ] #3
- [ ] create server and client gramine enclave - use a Makefile, you can checkout examples [here](https://github.com/gramineproject/gramine/tree/master/CI-Examples) and also (geth-sgx-gramine)(https://github.com/flashbots/geth-sgx-gramine/blob/attestation_ratls/Makefile) is a good source of inspiration
- [ ] create ratls certificate on startup of server and client, you can reuse part of [geth-sgx-gramine's init script](https://github.com/flashbots/geth-sgx-gramine/blob/attestation_ratls/geth_init.cpp)

### PR DESCRIPTION
**Pull request by lonerapier**
Closes #1

### CODE REVIEW

**Comment by fnerdman** [comment]
Can you add certificate creation in both client and server? You could use part of the `geth_init.cpp` script. It would be even nicer if you would implement the certificate creation function in go via the ratls wrapper module.

**Comment by lonerapier** [comment]
> It would be even nicer if you would implement the certificate creation function in go via the ratls wrapper module.

Thanks. That's a nice suggestion. Will add it in the wrapper module.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+Run `make` (non-debug) or `make DEBUG=1` (debug) in the [server](./server/) and [client](./client/) directory.
+
+### Building for SGX
+
+Run `make SGX=1` (non-debug) or `make SGX=1 DEBUG=1` (debug) in the directory.
+
+### Enclave size
+
+To change the amount of memory allocated to the enclave to 8 GB, run `make SGX=1 ENCLAVE_SIZE=8G`. Default is 8 GB.
```
Comment (fnerdman) [inline README.md]: I think we can lower the amount of memory required, have you tried this?

File: server/server.go [hunk 2]
```diff
@@ -17,43 +21,40 @@ func main() {
...
-				KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
-				Roots: caCertPool,
-			}
-			for _, cert := range cs.PeerCertificates[1:] {
-				opts.Intermediates.AddCert(cert)
-			}
-			_, err := cs.PeerCertificates[0].Verify(opts)
-			*/
-			err = ra_tls_verify(cs.PeerCertificates[0].Raw)
+			err = ratls_wrapper.RATLSVerifyDer(cs.PeerCertificates[0].Raw, nil, nil, nil, nil)
```
Comment (fnerdman) [inline server/server.go]: The Client MRENCLAVE should be passed to the server via env variable and should be verified here. We should also add the MRSIGNER as a possible env variable. If the env variables are empty, the arguments should be `nil`

Comment (fnerdman) [reply server/server.go]: Vice versa for client.go

File: server/server.manifest.template [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+]
+
+sgx.allowed_files = [
+  "file:/etc/nsswitch.conf",
+  "file:/etc/localtime",
+  "file:/etc/hosts",
+  "file:/etc/ld.so.cache",
+  "file:/etc/host.conf",
+  "file:/etc/passwd",
+  "file:tls/tlskey.der",
```
Comment (fnerdman) [inline server/server.manifest.template]: With this we would be writing the tls key and cert unencrypted to the hostfs. Ideally the tls keys are saved to memory only via gramine tmpfs. The tls ket and cert locations `/tmp/tlskey.der` and  `/tmp/tlscert.der` can be left hardcoded and there's no need to configure them via env vars.

Comment (fnerdman) [reply server/server.manifest.template]: Vice versa with `client`

File: server/tls/tlscert.der
Comment (fnerdman) [inline server/tls/tlscert.der:1]: We should delete this file, as it is created in the enclave.

File: server/tls/tlskey.der
Comment (fnerdman) [inline server/tls/tlskey.der:1]: We should delete this file, as it is created in the enclave.

File: client/tls/tlscert.der
Comment (fnerdman) [inline client/tls/tlscert.der:1]: We should delete this file, as it is created in the enclave.

File: client/tls/tlskey.der
Comment (fnerdman) [inline client/tls/tlskey.der:1]: We should delete this file, as it is created in the enclave.

**Review by fnerdman — CHANGES_REQUESTED** [review]

File: server/server.go (hunk 2 above)
Comment (lonerapier) [reply server/server.go]: Understood, adding all four measurement args.

File: client/client.manifest.template [hunk 4]
```diff
@@ -21,7 +21,12 @@ loader.log_level = "{{ log_level }}"
 # applies. Paths must be in-Gramine visible paths, not host paths (i.e.,
 # paths must be taken from fs.mounts[...].path, not fs.mounts[...].uri).
 loader.env.LD_LIBRARY_PATH = "/lib:{{ arch_libdir }}:/usr/lib:/usr/{{ arch_libdir }}"
-loader.env.DEBUG = "1"
+
+loader.env.DEBUG = { passthrough = true }
+loader.env.MRENCLAVE = { passthrough = true }
+loader.env.MRSIGNER = { passthrough = true }
+loader.env.ISV_PROD_ID = "{{ isvprodid }}"
+loader.env.ISV_SVN = "{{ isvsvn }}"
```
Comment (fnerdman) [inline client/client.manifest.template]: I think we should pass these through as well. What is happening here is, we provide the isvprodid and isvsvn of the client enclave to use these values to attest the report of the server enclave, which potentially may have different values.

File: client/client.go [hunk 5]
```diff
@@ -17,40 +15,40 @@ import (
...
 	// Read the key pair to create certificate
-	cert, err := mutual_ratls.LoadX509KeyPairDER(tlsCertPath, tlsKeyPath)
+	cert, err := mutual_ratls.X509KeyPairDER(derKey, derCert)
 	if err != nil {
 		log.Fatal(err)
 	}
 
+	mrenclave := mutual_ratls.GetSGXEnvVar("MRENCLAVE")
+	if mrenclave == nil {
+		log.Fatal("required mrenclave for enclave measurement.")
```
Comment (fnerdman) [inline client/client.go]: Why fail if mrenclave is nil?

File: server/server.go [hunk 6]
```diff
@@ -21,37 +19,36 @@ func main() {
...
-	derCrt, err := mutual_ratls.LoadX509KeyPairDER(tlsCertPath, tlsKeyPath)
+	cert, err := mutual_ratls.X509KeyPairDER(derKey, derCrt)
 	if err != nil {
-		log.Fatal(err)
+		log.Fatal(err.Error())
 	}
 
+	mrenclave := mutual_ratls.GetSGXEnvVar("MRENCLAVE")
+	if mrenclave == nil {
+		log.Fatal("required mrenclave for enclave measurement")
```
Comment (fnerdman) [inline server/server.go]: Same as with client.

File: server/server.manifest.template [hunk 7]
```diff
@@ -21,7 +21,12 @@ loader.log_level = "{{ log_level }}"
 # applies. Paths must be in-Gramine visible paths, not host paths (i.e.,
 # paths must be taken from fs.mounts[...].path, not fs.mounts[...].uri).
 loader.env.LD_LIBRARY_PATH = "/lib:{{ arch_libdir }}:/usr/lib:/usr/{{ arch_libdir }}"
-loader.env.DEBUG = "1"
+
+loader.env.DEBUG = { passthrough = true }
+loader.env.MRENCLAVE = { passthrough = true }
+loader.env.MRSIGNER = { passthrough = true }
+loader.env.ISV_PROD_ID = "{{ isvprodid }}"
+loader.env.ISV_SVN = "{{ isvsvn }}"
```
Comment (fnerdman) [inline server/server.manifest.template]: Same as with client.

**Review by fnerdman — CHANGES_REQUESTED** [review]

File: client/client.manifest.template (hunk 4 above)
Comment (lonerapier) [reply client/client.manifest.template]: Oh you're right 👍 , changing this.

File: client/client.go (hunk 5 above)
Comment (lonerapier) [reply client/client.go]: I misunderstood that verifying mrenclave is required. will change this to optional.

**Review by fnerdman — APPROVED** [review]

**Pull request merged**